src/nupic/algorithms/Connections.cpp

```diff
@@ -418,78 +389,35 @@ bool segmentIncreasingOrder(const SegmentOverlap& a, const SegmentOverlap& b)
   return a.segment < b.segment;
 }
 
-void Connections::computeActivity(const vector<Cell>& input,
+void Connections::computeActivity(const vector<CellIdx>& input,
                                   Permanence activePermanenceThreshold,
                                   SynapseIdx activeSynapseThreshold,
                                   Permanence matchingPermanenceThreshold,
                                   SynapseIdx matchingSynapseThreshold,
-                                  vector<SegmentOverlap>& outActiveSegments,
-                                  vector<SegmentOverlap>& outMatchingSegments,
-                                  bool recordIteration)
+                                  vector<SegmentOverlap>& activeSegmentsOut,
+                                  vector<SegmentOverlap>& matchingSegmentsOut)
+  const
 {
-  vector<UInt32> numActiveSynapsesForSegment(nextFlatIdx_, 0);
-  vector<UInt32> numMatchingSynapsesForSegment(nextFlatIdx_, 0);
+  SegmentExcitationTally excitations(*this, activePermanenceThreshold,
```

Comment:
I profiled it before committing the change. I didn't see a perf impact. I figure if we start to see a perf impact, we can add new methods that are specific to data structures.

Yeah, I thought deeply about iterators. They'll work for single data structures, but they won't work for the general case, e.g. the case in this PR: two vectors. There's really not a way to create an iterator that traverses two vectors. Boost offers this functionality, but the iterator has the perf overhead of having to check whether it has reached the end of the first vector on every ++.